libs/server/Storage/Session/MainStore/MainStoreOps.cs

```diff
@@ -597,43 +597,54 @@ private unsafe GarnetStatus RENAME(ArgSlice oldKeySlice, ArgSlice newKeySlice, S
-                        ((RespInputHeader*)inputPtr)->cmd = RespCommand.GETWITHETAG;
-                        ((RespInputHeader*)inputPtr)->flags = 0;
 
-                        // Check if etag is nil for getWithEtagStatus, this tells us if the key already had an etag associated with it
                         bool hasEtag = false;
-                        SpanByteAndMemory etagAndDataOutput = new SpanByteAndMemory();
-                        GarnetStatus getWithEtagStatus = GET(ref oldKey, ref getWithEtagInput, ref etagAndDataOutput, ref context);
-                        // we know this key exists, so there is no reason for the status to not be OK
-                        Debug.Assert(getWithEtagStatus == GarnetStatus.OK);
+
                         // since we didn't give the etagAndDataOutput span any memory when creating it, the backend would necessarily have had to allocate heap memory if item is not NOTFOUND
                         Debug.Assert(!etagAndDataOutput.IsSpanByte);
-                        MemoryHandle outputMemHandle = etagAndDataOutput.Memory.Memory.Pin();
+                        using MemoryHandle outputMemHandle = etagAndDataOutput.Memory.Memory.Pin();
                         byte* outputBufCurr = (byte*)outputMemHandle.Pointer;
                         byte* end = outputBufCurr + etagAndDataOutput.Length;
+
+                        // GETWITHETAG returns an array of two items, etag, and value.
+                        // we need to read past RESP metadata and control sequences to get to etag, and value
                         RespReadUtils.ReadUnsignedArrayLength(out int numItemInArr, ref outputBufCurr, end);
-                        Debug.Assert(numItemInArr == 2);
-                        hasEtag = !RespReadUtils.ReadNil(ref outputBufCurr, end, out byte? _);
+
+                        Debug.Assert(numItemInArr == 2, "GETWITHETAG output RESP array should be of 2 elements only.");
+
+                        // we know a key-val pair does not have an etag if the first element is not null
+                        // if read nil is successful it will point to the ptrVal otherwise we need to re-read past the etag as RESP int64
+                        byte* startOfEtagPtr = outputBufCurr;
+                        hasEtag = !RespReadUtils.ReadNil(ref outputBufCurr, end, out _);
+                        if (hasEtag)
+                        {
+                            // read past the etag so we can then get to the ptrVal, we don't need specific val of etag, just need to move past it if it exists
+                            bool etagReadingSuccessful = RespReadUtils.Read64Int(out _, ref startOfEtagPtr, end);
+                            Debug.Assert(etagReadingSuccessful, "Etag should have been read succesffuly");
+                            // now startOfEtagPtr is past the etag and points to value 
+                            outputBufCurr = startOfEtagPtr;
```

Comment:
Why not simply pass "ref outputBufCurr" on Read64Int? It'd also leave startOfETagPtr where it would be, which would reduce surprise if it's used elsewhere. (It's not, here, and in some cases I use {} to create a temp scope just to have these very-local variables scoped as minimally as possible)

Reply:
refactored this code entirely